crates/core/src/rlp/decode.rs

```diff
@@ -49,174 +67,188 @@ impl RLPDecode for u8 {
 
 impl RLPDecode for ethereum_types::H32 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H32)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H32(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::H64 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H64)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H64(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::H128 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H128)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H128(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::H256 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H256)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H256(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::H264 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H264)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H264(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::Address {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H160)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H160(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::H512 {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H512)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H512(value), rest))
     }
 }
 
 impl RLPDecode for ethereum_types::Signature {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
-        RLPDecode::decode(rlp).map(ethereum_types::H520)
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
+        let (value, rest) = RLPDecode::decode_unfinished(rlp)?;
+        Ok((ethereum_types::H520(value), rest))
+    }
+}
+
+impl RLPDecode for ethereum_types::U256 {
```

Comment:
I think we should use `crate::U256` here instead. Same for other `ethereum_types`